File: Homework/Hoang_Nguyen/lesson 1/202_happy_number.js
/**
 * @param {number} n
 * @return {boolean}
 */
var isHappy = function (n) {
  function nToSum(n) {
    let sum = 0;
    while (n) {
      sum += Math.pow(n % 10, 2);
      n = Math.floor(n / 10);
    }
    return sum;
  }

  let slow = n;
  let fast = n;

  while (fast !== 1) {
    slow = nToSum(slow);
    fast = nToSum(fast);
    if (fast === 1) {
      break;
    }
    fast = nToSum(fast);

    if (slow === fast) {
      return false;
    }
  }
  return true;
};

/**
 * @param {number} n
 * @return {boolean}
 */
var isHappyBySet = function (n) {
  let set = new Set();
  while (!set.has(n)) {
    set.add(n);
    let sum = 0;
    while (n) {
      sum += Math.pow(n % 10, 2);
      n = Math.floor(n / 10);
    }

    if (sum === 1) {
      return true;
    }
    n = sum;
  }
  return false;
};

File: Homework/quoc_khanh/lesson_1/202.HappyNumber.py
from typing import List


def getDigits(n: int) -> List[int]:
    if n == 0:
        return [0]
    digits = []
    while n > 0:
        digit = n % 10
        digits.append(digit)
        n = n // 10
    return digits

def getSumSquareOfDigits(n: int) -> int:
    sum = 0
    digits = getDigits(n)
    for digit in digits:
        sum += digit ** 2
    return sum

class Solution:
    def isHappy(self, n: int) -> bool:
        slow, fast = n, n
        while True:
            slow = getSumSquareOfDigits(slow)
            fast = getSumSquareOfDigits(fast)
            fast = getSumSquareOfDigits(fast)
            if slow == 1 or fast == 1:
                return True
            if slow == fast:
                return False
            